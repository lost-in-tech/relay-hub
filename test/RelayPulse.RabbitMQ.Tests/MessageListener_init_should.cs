using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.DependencyInjection.Extensions;
using RelayPulse.Core;
using RelayPulse.RabbitMQ.Subscribers;
using RelayPulse.RabbitMQ.Tests.Fakes;
using RelayPulse.RabbitMQ.Tests.Helpers;
using Shouldly;

namespace RelayPulse.RabbitMQ.Tests;

public class MessageListener_init_should(IocFixture fixture) : IClassFixture<IocFixture>
{
    [Fact]
    public async Task Setup_exchange_and_queues_when_queue_settings_empty()
    {
        var sp = fixture.Services();
        var sut = sp.GetRequiredService<ISetupRabbitMq>();
        await Should.ThrowAsync<RelayPulseException>(async () =>  await sut.Run(CancellationToken.None));
    }

    [Theory]
    [MemberData(nameof(TestData))]
    public async Task Setup_exchange_and_queues(TestInput<RabbitMqSettings> input)
    {
        var sp = fixture.Services(sc => { sc.Replace(ServiceDescriptor.Singleton<IQueueSettings>(input.Data)); });
        var sut = sp.GetRequiredService<ISetupRabbitMq>();
        await sut.Run(CancellationToken.None);
        var gotWrapper = sp.GetRequiredService<IRabbitMqWrapper, FakeRabbitMqWrapper>();

        new
        {
            gotWrapper.ExchangeDeclareCalls,
            gotWrapper.QueueDeclareCalls,
            gotWrapper.QueueBindCalls
        }.ShouldMatchContent(input.Key, input.Scenario);
    }

    public static IEnumerable<object[]> TestData = new TestInput<RabbitMqSettings>[]
    {
        new()
        {
            Key = "1",
            Scenario = "When default exchange is fanout",
            Data = new RabbitMqSettings
            {
                DefaultExchange = "default-exchange",
                DefaultExchangeType = ExchangeTypesSupported.Fanout,
                Queues =
                [
                    new QueueSettings
                    {
                        Name = "queue-fanout",
                    }
                ]
            }
        },
        new()
        {
            Key = "2",
            Scenario = "When exchange is fanout and name provided as part of queue",
            Data = new RabbitMqSettings
            {
                DefaultExchange = "default-exchange",
                DefaultExchangeType = ExchangeTypesSupported.Headers,
                Queues =
                [
                    new QueueSettings
                    {
                        Exchange = "non-default-exchange",
                        ExchangeType = ExchangeTypesSupported.Fanout,
                        Name = "queue-fanout"
                    }
                ]
            }
        },
        new()
        {
            Key = "3",
            Scenario = "When exchange is topic and name provided as part of queue",
            Data = new RabbitMqSettings
            {
                Queues =
                [
                    new QueueSettings
                    {
                        Exchange = "non-default-exchange",
                        ExchangeType = ExchangeTypesSupported.Topic,
                        Name = "queue-topic",
                        Bindings =
                        [
                            new QueueBinding
                            {
                                RoutingKey = "test-route-key"
                            }
                        ]
                    }
                ]
            }
        },
        new()
        {
            Key = "4",
            Scenario = "When exchange is header and name provided as part of queue",
            Data = new RabbitMqSettings
            {
                Queues =
                [
                    new QueueSettings
                    {
                        Exchange = "non-default-exchange",
                        ExchangeType = ExchangeTypesSupported.Headers,
                        Name = "queue-header",
                        Bindings =
                        [
                            new QueueBinding
                            {
                                MatchAny = true,
                                Headers = new Dictionary<string, string>{
                                    ["event-name"] = "order-created"
                                }
                            }
                        ]
                    }
                ]
            }
        },
        new()
        {
            Key = "5",
            Scenario = "a real world example",
            Data = new RabbitMqSettings
            {
                DefaultExchange = "bookworm.orders",
                DefaultExchangeType = ExchangeTypesSupported.Topic,
                Queues = [
                    new QueueSettings
                    {
                        Name = "bookworm.order-created.email",
                        Bindings = new []
                        {
                            new QueueBinding
                            {
                                RoutingKey = "Bookworm.OrderCreated"
                            }
                        }
                    },
                    new QueueSettings
                    {
                        Name = "bookworm.order-created.slack",
                        Bindings = new []
                        {
                            new QueueBinding
                            {
                                RoutingKey = "Bookworm.OrderCreated"
                            }
                        }
                    }
                ]
            }
        }
    }.ToTestData();
}