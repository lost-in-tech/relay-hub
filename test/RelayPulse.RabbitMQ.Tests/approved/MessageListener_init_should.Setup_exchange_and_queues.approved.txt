{
  "exchangeDeclareCalls": [
    {
      "name": "default-exchange",
      "type": "direct"
    },
    {
      "name": "default-exchange-dlx",
      "type": "topic"
    },
    {
      "name": "default-exchange-rtx",
      "type": "topic"
    }
  ],
  "queueDeclareCalls": [
    {
      "name": "default-exchange-dlq"
    },
    {
      "name": "default-exchange-rtq",
      "args": {
        "x-dead-letter-exchange": "default-exchange",
        "x-message-ttl": 60000
      }
    },
    {
      "name": "default-queue",
      "args": {
        "x-dead-letter-exchange": "default-exchange-dlx"
      }
    },
    {
      "name": "default-exchange-dlq"
    },
    {
      "name": "default-exchange-rtq",
      "args": {
        "x-dead-letter-exchange": "default-exchange",
        "x-message-ttl": 60000
      }
    },
    {
      "name": "default-queue2",
      "args": {
        "x-dead-letter-exchange": "default-exchange-dlx"
      }
    }
  ],
  "queueBindCalls": [
    {
      "queue": "default-exchange-dlq",
      "exchange": "default-exchange-dlx",
      "routeKey": "*"
    },
    {
      "queue": "default-exchange-rtq",
      "exchange": "default-exchange-rtx",
      "routeKey": "*"
    },
    {
      "queue": "default-queue",
      "exchange": "default-exchange",
      "routeKey": "rp:target:default-queue"
    },
    {
      "queue": "default-queue",
      "exchange": "default-exchange",
      "routeKey": ""
    },
    {
      "queue": "default-exchange-dlq",
      "exchange": "default-exchange-dlx",
      "routeKey": "*"
    },
    {
      "queue": "default-exchange-rtq",
      "exchange": "default-exchange-rtx",
      "routeKey": "*"
    },
    {
      "queue": "default-queue2",
      "exchange": "default-exchange",
      "routeKey": "rp:target:default-queue2"
    },
    {
      "queue": "default-queue2",
      "exchange": "default-exchange",
      "routeKey": ""
    }
  ]
}